### PR TITLE
ui: keep table components from being destroyed when viewing details

### LINKED ISSUE #641: Slow log search result should be cached
## Bug Report

Please answer these questions before submitting your issue. Thanks!

**What did you do?**

Search slow log, and click one to see the details. After moving back, it starts search again.

**What did you expect to see?**

Display the result immediately

**What did you see instead?**

Nothing shows up, and the refresh button begin spinning.

**What version of TiDB Dashboard are you using (`./tidb-dashboard --version`)?**

I'm using v4.0.0 GA, not sure what version is Dashboard.

### PR DESCRIPTION
**Pull request by unbyte**
preview:

known issue: 
- meaningless transition animation

close #641

### CODE REVIEW

File: ui/lib/components/ListAndDetail/index.tsx [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    >
+      <ListComponent />
+      {detailPathMatcher(location.pathname) && (
+        <DetailComponent
+          style={{
+            zIndex: 99,
+            position: 'absolute',
+            top: 0,
+            left: 0,
+            background: '#fff',
```
Comment (unbyte) [inline ui/lib/components/ListAndDetail/index.tsx:31]: does the background color need to be determined by theme?

**Comment by baurine** [comment]
I will have a look later today

File: ui/lib/components/ListAndDetail/index.tsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  detailPathMatcher: (path: string) => boolean
+}
+
+export default function ({
+  ListComponent,
+  DetailComponent,
+  detailPathMatcher,
+}: ListAndDetailProps) {
+  const location = useLocation()
+  console.log('rendered')
```
Comment (baurine) [inline ui/lib/components/ListAndDetail/index.tsx:16]: It can be removed.

File: ui/lib/components/ListAndDetail/index.tsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  return (
+    <div
+      style={{
+        position: 'relative',
+      }}
+    >
+      <ListComponent />
+      {detailPathMatcher(location.pathname) && (
+        <DetailComponent
```
Comment (baurine) [inline ui/lib/components/ListAndDetail/index.tsx:25]: How about use an extra `div` to wrap the `<DetailComponent/>`? for example:

```
<div className={xxx}>
  <DetailComponent/>
</div>
```

So we don't need to modify the DetailPage.

**Comment by baurine** [comment]
Although show List and Detail inside one page is a kind of solution, I am more feel it as tricky. How about importing a small store (only stores the list data) for each of these 2 apps? Or, cache the list data in the useStatementTableController.ts and useSlowQueryTableController.ts. I prefer the latter.

For example:

```js
// useStatementTableController.ts

const cachedListData = {
  key: '',  // beginTime_endTime_fields_schemas_stmtTypes_searchText_refreshTimes
  listData: [] // StatementModel[]
}

export default function useStatementTableController() {
  //...

  useEffect(() => {
    async function queryStatementList() {
      // check cache
      const curKey = `${validTimeRange.begin_time}_${validTimeRange.end_time!}..._${refreshTimes}`
      if (curKey === cachedListData.key) {
        return cachedListData.listData
      }
     
      //...
    }

    queryStatementList()
  }, [queryOptions, allTimeRanges, validTimeRange, selectedFields])

```

How do you think about it?

**Comment by unbyte** [comment]
> How about importing a small store (only stores the list data) for each of these 2 apps

Before this PR, I have made many attempts, like implementing component caching like https://github.com/CJY0208/react-activation, or implementing a simple `useCache` hook, judge whether there is a cached state when the component is mounted and use the cache state to render the component.

One disadvantage of the above is the introduction of a relatively large amount of code. The first one is self-evident. The second one needs to maintain a state machine of route inside cache store and modify the logic in the controllers for using cached state and caching state.

In fact, the solution in this PR is simple enough and has no performance issues, though the transition animation makes the appearance of the details page a little less smooth.

File: ui/lib/components/ListAndDetail/index.tsx (hunk 3 above)
Comment (unbyte) [reply ui/lib/components/ListAndDetail/index.tsx:25]: Just worried that the introduction of a new div will lead to the failure of some styles related to the div level. And it seems there is no such problem at present,

File: ui/lib/components/ListAndDetail/index.tsx (hunk 2 above)
Comment (unbyte) [reply ui/lib/components/ListAndDetail/index.tsx:16]: my fault😂

File: ui/lib/apps/SlowQuery/index.tsx [hunk 4]
```diff
@@ -1,16 +1,18 @@
...
     <Root>
       <Router>
-        <Routes>
-          <Route path="/slow_query" element={<List />} />
-          <Route path="/slow_query/detail" element={<Detail />} />
-        </Routes>
+        <ListAndDetail
+          DetailComponent={Detail}
+          ListComponent={List}
+          detailPathMatcher={(path) => path.startsWith('/slow_query/detail')}
```
Comment (breeswish) [inline ui/lib/apps/SlowQuery/index.tsx:14]: This kind of manual path match is too hacky

**Comment by breeswish** [comment]
IMO, simple is good but should not sacrifice maintainability, i.e. it should not be a simple "hack". It should be a simple "solution".

**Pull request closed without merging**